File: de.vogella.scala.functions/src/de/vogella/scala/functions/Factorial.scala
package de.vogella.scala.functions

object Factorial {
  
 def main(args: Array[String]) {
   val f = Factorial
   println( f.factorial(5) )
   println( f.factorial(6) )
   println( f.factorial(7) )
   println( f.factorial(8) )
 }
 
 def factorial (x:BigInt):BigInt= {
   if (x==0) 1
   else x* factorial (x-1)
 }
}

File: de.vogella.algorithms.riddles/src/de/vogella/recursion/Factorial.java
package de.vogella.recursion;

public class Factorial {
	public static long factor(long n) {
		// Base case
		if (n == 1) {
			return n;
		}
		// Reductions steps
		return n * factor(n - 1);
	}

	public static void main(String[] args) {
		assert (factor(5) == 120);
		assert (factor(6) == 720);
	}
}
